pkg/agent/multicast/mcast_discovery.go

```diff
@@ -97,14 +110,83 @@ func (s *IGMPSnooper) queryIGMP(group net.IP, versions []uint8) error {
+		// from table0. The final OpenFlow message will use a standard OpenFlow port number OFPP_TABLE = 0xfffffff9 corrected
+		// by ofnet.
+		outPort := uint32(0)
+		if err := s.ofClient.SendIGMPQueryPacketOut(igmpQueryDstMac, types.McastAllHosts, outPort, igmp); err != nil {
 			return err
 		}
-		klog.V(2).InfoS("Sent packetOut for IGMP query", "group", group.String(), "version", version)
+		klog.V(2).InfoS("Sent packetOut for IGMP query", "group", group.String(), "version", version, "outPort", outPort)
 	}
 	return nil
 }
 
+func (s *IGMPSnooper) validate(event *mcastGroupEvent, packetInData protocol.Ethernet) (bool, error) {
+	if s.validator == nil {
+		// Do not need to validate if the packet should be dropped since Validator is null, thus here returns directly
+		return true, nil
+	}
+	if event.iface.Type != interfacestore.ContainerInterface {
+		return true, fmt.Errorf("interface is not container")
+	}
+	// Validates check if packet should be dropped or not, and return multicast NP information
+	item, err := s.validator.Validate(event.iface.PodName, event.iface.PodNamespace, event.group)
+	if err != nil {
+		// It shall drop the packet if function Validate returns error
+		klog.ErrorS(err, "Function Validate returned")
+		return false, err
+	}
+	if item.NPType != nil {
+		klog.V(2).InfoS("Got NetworkPolicy action for IGMP query", "RuleAction", *item.RuleAction,
+			"UUID", item.UUID, "Name", item.Name)
+		s.addToIGMPReportNPStatsMap(item, uint64(packetInData.Len()))
+	}
+	if item.RuleAction != nil && *item.RuleAction == v1alpha1.RuleActionDrop {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (s *IGMPSnooper) addToIGMPReportNPStatsMap(item types.McastNPValidationItem, packetLen uint64) {
+	updateRuleStats := func(igmpReportStatsMap map[apitypes.UID]map[string]*types.RuleMetric, uuid apitypes.UID, name string) {
+		if igmpReportStatsMap[uuid] != nil {
+			if igmpReportStatsMap[uuid][name] != nil {
+				stats := igmpReportStatsMap[uuid][name]
+				newStats := types.RuleMetric{Bytes: stats.Bytes + packetLen, Packets: stats.Packets + 1}
+				igmpReportStatsMap[uuid][name] = &newStats
+			} else {
+				igmpReportStatsMap[uuid][name] = &types.RuleMetric{Bytes: packetLen, Packets: 1}
+			}
+		} else {
+			igmpReportStatsMap[uuid] = make(map[string]*types.RuleMetric)
+			igmpReportStatsMap[uuid][name] = &types.RuleMetric{Bytes: packetLen, Packets: 1}
+		}
+	}
+	ruleType := *item.NPType
+	if ruleType == v1beta2.AntreaNetworkPolicy {
+		s.igmpReportANPStatsMutex.Lock()
+		updateRuleStats(s.igmpReportANPStats, item.UUID, item.Name)
+		s.igmpReportANPStatsMutex.Unlock()
+	} else if ruleType == v1beta2.AntreaClusterNetworkPolicy {
+		s.igmpReportACNPStatsMutex.Lock()
+		updateRuleStats(s.igmpReportACNPStats, item.UUID, item.Name)
+		s.igmpReportACNPStatsMutex.Unlock()
+	}
+}
+
+func (s *IGMPSnooper) collectStats() (igmpANPStats, igmpACNPStats map[apitypes.UID]map[string]*types.RuleMetric) {
```

Comment:
Please add comments say: This func will reset the saved stats.

Reply:
updated